### DISCUSSION TITLE
Thảo luận về xuất bản pdf

### DISCUSSION THREAD

**Discussion by thanhcsf** [d2l-ai/d2l-vi#3095] [General]
Anh em trao đổi thêm ở đây về vấn đề này nhé. 
CC: @quangnhat185 @nguyenhoa93

**Comment by lkhphuc** [comment]
Nếu được mọi người cho mình xin cái tóm tắt về việc này được không. Trước giờ mình tưởng dịch xong có html rồi thì cứ Crt+P "Print as PDF" là xong chứ 😅

**Reply by thanhcsf** [reply]
Nó có format theo bản gốc đấy ông giáo, như hồi xưa nhóm mình làm MLY thì cũng có script để export ra cho nhanh và custom được nhiều hơn. Bản này đỡ cái là nhóm tác giả viết cho sẵn rồi, có command line để export PDF rồi, nhưng mà vẫn có khả năng xảy ra lỗi nên vẫn cần người handle phần này 😄

**Reply by quangnhat185** [reply]
@lkhphuc Team mình muốn làm file pdf xịn xò như [bản pdf tiếng anh](https://d2l.ai/d2l-en.pdf) và có thể tự generate khi có thay đổi trong nội dung luôn.

**Reply by lkhphuc** [reply]
Oh nhìn rất fancy.

**Comment by quangnhat185** [comment]
Hiện tại theo [hướng dẫn](http://book.d2l.ai/user/deploy.html#deploying-through-github) của bản d2l gốc thì việc deploy pdf file khá đơn giản, chỉ cần sử dụng: 
`d2lbook deploy html pdf`

Mình thấy trong file [deploy.sh](https://github.com/aivivn/d2l-vn/blob/master/deploy.sh) anh @tiepvupsu đã dùng cú pháp `d2lbook deploy html` để chạy rồi. Tuy nhiên do mình chưa deploy cuốn này khi nào nên chưa rõ cách build để thử cú pháp tạo pdf trên, @nguyenhoa93  có thể giúp được không?

**Reply by tiepvupsu** [reply]
À, mình không @quangnhat185 
`d2lbook build html; d2lbook deploy html` để build và deploy html lên web.

Anh đoán để buils pdf, ta dùng lệnh
`d2lbook build pdf`.

**Comment by hnguyentt** [comment]
Mình build thử rồi: https://drive.google.com/file/d/1C7PS-Ynu4868abAvdN1VN1lZtd-yprOh/view?usp=sharing

Bị thiếu một số package nên hiện tại còn một số vấn đề:
1. Không hiển thị TOC
2. Không hiển thị tên Fig
3. Không hiển thị Ref

Mình chưa có thời gian để xem lại.

**Reply by thanhcsf** [reply]
Bản build này cho thấy nhiều vấn đề như:
- Ba vấn đề như trên.
- Chưa bỏ phần giới thiệu của nhóm dịch.
- Các phần cảm ơn người dịch của từng phần đều hiển thị.
- Fonts chữ chưa ổn cho lắm.
- Code khi sang trang có chú thích như (continued from previous page)
- Lỗi hiển thị bảng (trang 546)

Mình đang nghĩ không biết có thể tách phần build pdf này ra riêng, tách bạch với build html không; vì bản pdf này mình thấy cần customize nhiều hơn để phù hợp với bản tiếng Việt; trong khi đó bản web thì mình thấy nên bám sát với bản gốc để mình đỡ công.
@nguyenhoa93 cô giáo thấy việc này có khả thi không?

**Reply by hnguyentt** [reply]
Cái này chắc nhờ anh @tiepvupsu share folder `eval` trong build cho mọi người sử dụng để build `pdf` cho nhanh ạ.

Thật ra mình không thể chạy trực tiếp package `d2lbook` để generate `pdf` được vì mình thiếu fonts và rất nhiều dependent packages trong latex. Mình phải generate latex rồi từ file latex đó compile pdf, thiếu packages hay fonts nào thì phải add thêm packages đó.

**Reply by tiepvupsu** [reply]
@duythanhvn Sẽ có rất nhiều vấn đề nhỏ nữa, nhưng mình làm theo thứ tự quan trọng. Quan trọng nhất là giờ phải  hiển thị đủ các loại format (text, hình, công thức, reference). Sau đó mới đến hiển thị đúng, dàn trang, ...

@nguyenhoa93 Ok, anh sẽ shared folder eval.

**Reply by tiepvupsu** [reply]
P/S: Lo format trước, lo nội dung sau. Nội dung dù sao mình cũng phải chờ đến khi hoàn thiện html mới nên chỉnh sửa, nếu không giữ hai bản khác nhau sẽ rất khó bảo trì (maintain).

**Reply by tiepvupsu** [reply]
@nguyenhoa93 folder eval đây nhé https://drive.google.com/file/d/1BhB2vtkf-V-W4C64fz1MqIifywonC_xR/view?usp=sharing

**Reply by quangnhat185** [reply]
@nguyenhoa93 có thể liệt kê các bước bạn dùng để build file không? Cách sử dụng file `eval` này như thế nào?

Trường hợp thiếu font hay dependencies mình có thể thêm vào lúc build Dockerfile, bây giờ quan trọng là làm sao build được với format chuẩn.

**Comment by forummlcb** [comment]
Hi mọi  người, task này có tiến triển gì không?

**Comment by hnguyentt** [comment]
Mọi người vào check thử bản mới này. Có vẻ như giải quyết được các vấn đề trước đó!
https://drive.google.com/file/d/1C7PS-Ynu4868abAvdN1VN1lZtd-yprOh/view?usp=sharing
@duythanhvn @quangnhat185 

Bản này còn thiếu mấy chương sau. Nhờ anh @tiepvupsu up lại folder _build để em build lại bản pdf. Chỉ có folder eval thôi thì nó bị conflict nên em build ko được ạ.

**Reply by thanhcsf** [reply]
Xuất sắc cô giáo ơi, để tối mình xem kỹ. Nhờ cô giáo cho anh em biết quy trình để có được bản pdf này với nhé.

**Reply by quangnhat185** [reply]
Hay quá @nguyenhoa93 ơi, vậy là chỉ càn build lại là được phải không? Phần glossarry có hơi lỗi hiện thị tí xíu nhưng minor có thể sửa sau.

**Reply by hnguyentt** [reply]
1. Bước 1: install `d2lbook` theo hướng dẫn ở đây: https://github.com/d2l-ai/d2l-book/blob/master/docs/install.md
2. Download 3 font families (nếu máy bạn chưa có) & install. Nếu bạn dùng MacOSx thì cứ copy hết mấy font vào `/Library/Fonts`, fontspec sẽ tự detect.
* https://fonts.google.com/specimen/Source+Serif+Pro?subset=vietnamese&query=Source+Serif+
* https://fonts.google.com/specimen/Source+Sans+Pro?subset=vietnamese&query=Source+S
* https://fonts.google.com/specimen/Inconsolata?subset=vietnamese&query=inc
3. Chạy lệnh `d2lbook build pdf`. Sử dụng folder build mà anh Tiếp chi sẻ để ko cần phải eval lại.
4. Sau khi chạy thì sẽ thấy một đống lỗi xuất hiện. Gõ `X` để thoát.
5. `cd` vào `_build/pdf`, chạy các comment sau:
```
sed -i.bak 's/documentclass[letterpaper,11pt,english]{sphinxmanual}/documentclass[letterpaper,11pt,vietnamese]{sphinxmanual}/g' d2l-vn.tex
lf=$'\n'; sed -i.bak "s/\usepackage{cmap}/\usepackage[english,vietnamese]{babel}\\$lf \usepackage[utf8]{inputenc}\\$lf \usepackage{cmap}/g" d2l-vn.tex
lf=$'\n'; sed -i.bak "s/\setmainlanguage{english}/\setmainlanguage{english}\\$lf \setotherlanguages{vietnamese}//g" d2l-vn.tex
sed -i.bak 's/text{mất_mát}/text{mất\_mát}/g' d2l-vn.tex
sed -i.bak 's/textrm{diện_tích}/textrm{diện\_tích}/g' d2l-vn.tex
sed -i.bak 's/textrm{tuổi_đời}/textrm{tuổi\_đời}/g' d2l-vn.tex
```
6. Tiếp đó chạy: `xelatex -recorder "d2l-vn.tex"` thì thu được sản phẩm.

**Reply by thanhcsf** [reply]
> 2\. Chạy thì sẽ thấy một đống lỗi xuất hiện. Gõ `X` để thoát.

Cảm ơn cô giáo hướng dẫn siêu chi tiết :))

**Reply by hnguyentt** [reply]
Chắc phải chạy eval lại Glossary chứ bản html cũng bị lỗi như vậy cho bảng chữ S
![img](https://i.imgur.com/Pp0PDVL.png)

**Comment by quangnhat185** [comment]
Thanks to @nguyenhoa93 mà phần format chính vậy là khá ổn rồi. Bây giờ có lẽ nên bàn tiếp các vấn đề hiển thị và đề xuất phương án giải quyết nhỉ? Đến hiện tại thì có vẻ chúng ta đang có 3 vấn đề chính sau:

1. __Chưa bỏ phần giới thiệu của nhóm dịch.__
2. __Các phần cảm ơn người dịch của từng phần đều hiển thị.__
3. __Lỗi hiển thị tại chương glossary (tràn lề phần link tới discussion)__

Vấn đề thứ 1 thì mình nghĩ có nhất thiết phải bỏ lời giới thiệu không? Thực ra thì có hay không có lời giới thiệu cũng không quá ảnh hưởng tới nội dung chính của cuốn sách.

Vấn đề thứ 2 và thứ 3 thì có hai cách giải quyết. Cách thứ nhất là chỉnh sửa trực tiếp nội dung trong các file markdown (Các link disscusion thì có thể embedded trong từ [link](https://d2l.aivivn.com/glossary.html); tên contributors thì có thể được liệt kê vào một phần Acknowledgement ở cuối sách?), sau đó thì build và deploy như bình thường. Như vậy lúc này 2 bản online và pdf sẽ giống nhau.

Cách thứ hai là tạo một python script để sửa `tex file` trước khi build ra pdf. Cách này sẽ khá phức tạp nhưng không đụng chạm tới bản online trong khi bản pdf thì được customize tùy ý.

Mọi người cho ý kiến nhé @tiepvupsu @duythanhvn @nguyenhoa93 .

**Reply by thanhcsf** [reply]
Cảm ơn ông giáo @quangnhat185 đề cập chi tiết. Hướng của mình hiện tại là:
1. Mình cũng đồng tình việc giữ nguyên. Mình đã có cập nhật để phù hợp với tình trạng hiện tại của dự án.
2. Phần này thì như ông giáo đề cập, mình đang tiến hành cập nhật loại bỏ tất cả các phần này từ từng phần nội dung; chuyển người đóng góp này sang tập tin Acknowledgement ở cuối sách. Phần này mình đã và đang cập nhật.

Trong hình dung của mình thì mình muốn đi theo hướng thứ hai mà ông giáo đề cập, đầu tiên là đảm bảo bản này được cấu hình sát nhất với bản gốc, tiếp theo là việc deployment chắc chắn là phức tạp và customize dựa trên gói d2l-book từ bản gốc. 

Ở bản PDF này có các điểm mà ban đầu khi bắt đầu ta cũng đề cập như thay đổi từ Fig, từ Section,...để bản dịch xuyên suốt nhất. Mặc dù biết là sẽ phức tạp nhưng trên đường dài cập nhật thì mình muốn làm cách này.

**Comment by ngcthuong** [comment]
Cảm ơn các bạn trẻ :D 
1. Những quyển sách dịch thường có lời giới thiệu, nhưng không nhất thiết. Nhưng mình nghĩ nên để
2. Nếu không muốn hiển thị người dịch từng chương có thể làm tổng hợp lại như dạng glossary với từng người đóng góp những chương nào. Có vẻ giống ý của @quangnhat185

**Comment by tiepvupsu** [comment]
Cảm ơn các bạn trẻ. Mình thiên về hướng viết một python script để sửa file tex. Cách này giúp không phải cập nhật bằng tay mỗi khi các file markdown có cập nhật mới và cũng không phải sửa các file markdown hiện tại.

Các chỗ như `Fig, Section, ...` cũng có thể thay trong các file build ra tiếng Việt được. Việc  này cũng có thể làm đồng thời cho bản trên web và deploy lại.

@quangnhat185 và @nguyenhoa93 có thể làm một PR đơn giản cho việc chạy file python script đó. Anh sau đó có thể giúp code những phần phức tạp (anh rất tin tưởng coding skill của @nguyenhoa93 nên chắc cũng không đến lượt).

**Reply by hnguyentt** [reply]
Dạ tuần sau em thử. Tuần này em có cái deadline quan trọng ạ!
